This patch adds a new "pickup_thrown" option to the game.  If pickup_thrown
and autopickup are both on, you'll automatically pick up things you threw,
even if you wouldn't normally have picked up that sort of thing.  It makes
using missile weapons less hassle and more fun.

This does change the game mechanics a bit, but it's not terribly different
from what you can already do by manipulating your pickup_types.  Without
this, if you're in a tight spot and you need the missiles you've already
thrown, you can temporarily add ) to your pickup types.  With pickup_thrown
you just don't have to worry about it.  I think it's a good tradeoff.

This patch was built on patchable Nethack, so you have to apply that
patch first.  It's at http://www.argon.org/~roderick/nethack/#patchable.

Bones/save compatibility:  This patch uses 1 object bit in a bones-safe
manner.  As long as you don't add more than 6 object bits this patch is
entirely bones compatible, and adding the patch is save compatible.
See http://www.argon.org/~roderick/nethack/#object-bits for details.

This patch is available at http://www.argon.org/~roderick/nethack/.

Roderick Schertler <roderick@argon.org>


diff -r -X /home/roderick/.diff-exclude -uN base.patchable/dat/opthelp work.pickup_thrown/dat/opthelp
--- base.patchable/dat/opthelp	2003-02-23 09:43:16.000000000 -0500
+++ work.pickup_thrown/dat/opthelp	2003-03-18 21:32:18.000000000 -0500
@@ -25,6 +25,7 @@
                delay code) if moving objects seem to teleport across rooms
 number_pad     use the number keys to move instead of yuhjklbn    [FALSE]
 perm_invent    keep inventory in a permanent window               [FALSE]
+pickup_thrown  autopickup things you threw                        [TRUE]
 prayconfirm    use confirmation prompt when #pray command issued  [TRUE]
 pushweapon     when wielding a new weapon, put your previously
                wielded weapon into the secondary weapon slot      [FALSE]
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/doc/Guidebook.mn work.pickup_thrown/doc/Guidebook.mn
--- base.patchable/doc/Guidebook.mn	2003-02-23 09:43:18.000000000 -0500
+++ work.pickup_thrown/doc/Guidebook.mn	2003-03-18 21:32:18.000000000 -0500
@@ -2040,6 +2040,12 @@
 level (Unburdened, Burdened, streSsed, straiNed, overTaxed,
 or overLoaded), you will be asked if you want to continue.
 (Default `S').
+.lp pickup_thrown
+If this boolean option is true and
+.op autopickup
+is on, try to pick up things that you threw, even if they aren't in
+.op pickup_types .
+Default is on.
 .lp pickup_types
 Specify the object types to be picked up when
 .op autopickup
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/doc/Guidebook.tex work.pickup_thrown/doc/Guidebook.tex
--- base.patchable/doc/Guidebook.tex	2003-02-23 09:43:18.000000000 -0500
+++ work.pickup_thrown/doc/Guidebook.tex	2003-03-18 21:32:18.000000000 -0500
@@ -2502,6 +2502,11 @@
 or overLoaded), you will be asked if you want to continue.
 (Default `S').
 %.lp
+\item[\ib{pickup\_thrown}]
+If this boolean option is true and {\it autopickup\/} is on, try to pick up
+things that you threw, even if they aren't in {\it pickup\_types\/}.
+Default is on.
+%.lp
 \item[\ib{pickup\_types}]
 Specify the object types to be picked up when ``{\it autopickup\/}'' 
 is on.  Default is all types.  If your copy of the game has the
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/include/flag.h work.pickup_thrown/include/flag.h
--- base.patchable/include/flag.h	2003-03-18 17:03:58.000000000 -0500
+++ work.pickup_thrown/include/flag.h	2003-03-18 21:32:18.000000000 -0500
@@ -63,6 +63,7 @@
 #endif
 	boolean  perm_invent;	/* keep full inventories up until dismissed */
 	boolean  pickup;	/* whether you pickup or move and look */
+	boolean  pickup_thrown;		/* auto-pickup items you threw */
 
 	boolean  pushweapon;	/* When wielding, push old weapon into second slot */
 	boolean  rest_on_space; /* space means rest */
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/include/obj.h work.pickup_thrown/include/obj.h
--- base.patchable/include/obj.h	2003-03-18 17:03:58.000000000 -0500
+++ work.pickup_thrown/include/obj.h	2003-03-18 21:32:18.000000000 -0500
@@ -91,7 +91,8 @@
 
 	Bitfield(in_use,1);	/* for magic items before useup items */
 	Bitfield(bypass,1);	/* mark this as an object to be skipped by bhito() */
-	/* 6 free bits */
+	Bitfield(was_thrown,1); /* thrown by the hero since last picked up */
+	/* 5 free bits */
 
 	int	corpsenm;	/* type of corpse is mons[corpsenm] */
 #define leashmon  corpsenm	/* gets m_id of attached pet */
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/src/bones.c work.pickup_thrown/src/bones.c
--- base.patchable/src/bones.c	2003-03-18 21:27:05.000000000 -0500
+++ work.pickup_thrown/src/bones.c	2003-03-18 21:33:16.000000000 -0500
@@ -82,6 +82,7 @@
 			otmp->rknown = 0;
 			otmp->invlet = 0;
 			otmp->no_charge = 0;
+			otmp->was_thrown = 0;
 
 			if (otmp->otyp == SLIME_MOLD) goodfruit(otmp->spe);
 #ifdef MAIL
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/src/dothrow.c work.pickup_thrown/src/dothrow.c
--- base.patchable/src/dothrow.c	2003-02-23 09:43:26.000000000 -0500
+++ work.pickup_thrown/src/dothrow.c	2003-03-18 21:32:18.000000000 -0500
@@ -871,6 +871,7 @@
 	boolean impaired = (Confusion || Stunned || Blind ||
 			   Hallucination || Fumbling);
 
+	obj->was_thrown = 1;
 	if ((obj->cursed || obj->greased) && (u.dx || u.dy) && !rn2(7)) {
 	    boolean slipok = TRUE;
 	    if (ammo_and_launcher(obj, uwep))
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/src/options.c work.pickup_thrown/src/options.c
--- base.patchable/src/options.c	2003-03-18 20:58:07.000000000 -0500
+++ work.pickup_thrown/src/options.c	2003-03-18 21:32:18.000000000 -0500
@@ -157,6 +157,7 @@
 	{"page_wait", (boolean *)0, FALSE, SET_IN_FILE},
 #endif
 	{"perm_invent", &flags.perm_invent, FALSE, SET_IN_GAME},
+	{"pickup_thrown", &flags.pickup_thrown, TRUE, SET_IN_GAME},
 	{"popup_dialog",  &iflags.wc_popup_dialog, FALSE, SET_IN_GAME},	/*WC*/
 	{"prayconfirm", &flags.prayconfirm, TRUE, SET_IN_GAME},
 	{"preload_tiles", &iflags.wc_preload_tiles, TRUE, DISP_IN_GAME},	/*WC*/
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/src/pickup.c work.pickup_thrown/src/pickup.c
--- base.patchable/src/pickup.c	2003-02-23 09:43:29.000000000 -0500
+++ work.pickup_thrown/src/pickup.c	2003-03-18 21:32:18.000000000 -0500
@@ -639,9 +639,11 @@
 
 
 #ifndef AUTOPICKUP_EXCEPTIONS
-	    if (!*otypes || index(otypes, curr->oclass))
+	    if (!*otypes || index(otypes, curr->oclass) ||
+		(flags.pickup_thrown && curr->was_thrown))
 #else
 	    if ((!*otypes || index(otypes, curr->oclass) ||
+		flags.pickup_thrown && curr->was_thrown ||
 		 is_autopickup_exception(curr, TRUE)) &&
 	    	 !is_autopickup_exception(curr, FALSE))
 #endif
@@ -651,9 +653,11 @@
 	    *pick_list = pi = (menu_item *) alloc(sizeof(menu_item) * n);
 	    for (n = 0, curr = olist; curr; curr = FOLLOW(curr, follow))
 #ifndef AUTOPICKUP_EXCEPTIONS
-		if (!*otypes || index(otypes, curr->oclass)) {
+		if (!*otypes || index(otypes, curr->oclass) ||
+			(flags.pickup_thrown && curr->was_thrown)) {
 #else
 	    if ((!*otypes || index(otypes, curr->oclass) ||
+			flags.pickup_thrown && curr->was_thrown ||
 		 is_autopickup_exception(curr, TRUE)) &&
 	    	 !is_autopickup_exception(curr, FALSE)) {
 #endif
@@ -1354,6 +1358,7 @@
 	    obj = splitobj(obj, count);
 
 	obj = pick_obj(obj);
+	obj->was_thrown = 0;
 
 	if (uwep && uwep == obj) mrg_to_wielded = TRUE;
 	nearload = near_capacity();
diff -r -X /home/roderick/.diff-exclude -uN base.patchable/util/makedefs.c work.pickup_thrown/util/makedefs.c
--- base.patchable/util/makedefs.c	2003-03-18 17:05:58.000000000 -0500
+++ work.pickup_thrown/util/makedefs.c	2003-03-18 21:32:18.000000000 -0500
@@ -718,6 +718,7 @@
 #  endif
 # endif
 #endif
+		"pickup thrown objects", /* EN "pickup thrown objects", */ // TODO DE
 #ifdef REDO
 		"redo command", /* EN "redo command", */ // TODO DE
 #endif
